### ISSUE TITLE
Apply deepwave to ultrasound

### ISSUE THREAD

**Issue by ruizhecao96** [ar4/deepwave#66]
Hello,
Thank you very much for this great library! I am curious if deepwave can be applied to ultrasound imaging. Ultrasound imaging usually uses much higher source frequency(e.g. 1MHz) and the imaging area is much smaller(e.g. 10cm^2). Is the acoustic equation or elastic wave equation implemented in deepwave still applicable in this situation? I would also like to ask if I can excite multiple sources simultaneously or with a certain time delay in deepwave?

**Comment by ar4** [comment]
Hello Ruizhe,

I am glad to hear that you are interested in Deepwave.

I think it should work for your situation, although you will need to
specify the "pml_freq" parameter to be equal to the approximate
frequency of your source when calling the propagator as the default
value for this (25Hz) is not appropriate and so the absorbing
boundaries would not work well.

Will you want to simulate an interface with air using the elastic
propagator (other than potentially having one along one or more edges
of the model)? If so, you might be able to get adequate results by
setting the shear modulus/v_s to zero in the air cells, but I suspect
that the accuracy might be low. Deepwave simply performs finite
difference wave propagation on the provided model - it doesn't have
special handling of acoustic-elastic interfaces (other than the option
of having free surfaces on the edges).

Exciting multiple sources in the same "shot" is easy to implement with
Deepwave. You specify the locations and source wavelets in arrays with
dimensions [number of shots, number of sources per shot, 2 (for
locations) / number of time samples (for wavelets)]. Your "number of
sources per shot" dimension will be of size greater than one.

Please let me know if you have any other questions or encounter any
difficulties.

**Comment by ruizhecao96** [comment]
Hello Alan,
Thank you very much for your detailed explanation! For now, I just use a scalar propagator to do some validation, and I did get promising results of wavefield! I will try later with an elastic propagator. 
However, when I want to set multiple sources locations I find the location parameter must be integer numbers. For example, I set a grid of 200x200 with griding space 0.1e^-3 m, and the source locations are [0, 0.5e-3, 1e-3, ...], only the first source location is valid. So the question is why the source locations should be long data type? Can I modify the source code to suit my case?

Best regards,
Ruizhe

**Comment by ar4** [comment]
Hello Ruizhe,

That is good news that the results are promising.

The source and receiver locations are provided as cell coordinates. If you
have a grid spacing of 0.1e-3m, and want a source at 0.5e-3m, then you
should specify its location as 5 to put it in the fifth cell. If, at some
point, you want a source or receiver at a location that is not a multiple
of the grid spacing, for example 0.05e-3m, then you can use multiple
sources/receivers to simulate it, as demonstrated in the location
interpolation example (
https://ausargeo.com/deepwave/example_location_interpolation.html).

**Comment by ruizhecao96** [comment]
Hello Alan,

Thank you for the suggestion! I figured out how to set the source locations properly. Specifically, I changed the indexing of the model in common.py in line 92.
original:
```
sa = source_amplitude[i]
sl = source_locations[i]
```
modified:
```
sa = source_amplitude[i]
sl = source_locations[i]
sl = (sl / grid_spacing).long()
```
In the function `pad_location`, line 432:
change `(locations + pad_starts_tensor).long()` to `(locations + pad_starts_tensor)`

Now I get the signal from the receivers, I found the amplitude of the signal remains consistent or even larger at the later time steps, I am curious if the scalar propagator considers the attenuation of the wave? If not, can I just multiply an exponential decay function with the received signal to simulate the attenuation?

Best regards
Ruizhe

**Comment by ar4** [comment]
Hello Ruizhe,

I am glad that you worked it out, however I suggest that it might be
easier to simply modify your calls to Deepwave rather than modifying
the code of Deepwave itself. For example, you could use

```python
deepwave.scalar(
    v, grid_spacing=dx, dt=dt,
    source_amplitudes=source_amplitudes,
    source_locations=(x_s/dx).long(),
    receiver_locations=(x_r/dx).long())
)
```

Waves in Deepwave will experience transmission loss and geometric
divergence, but it does not consider absorption. You can multiply the
receiver data output of Deepwave by an exponential if you need to
incorporate absorption and you think that implementing it that way
will be sufficiently accurate for your purposes.

**Comment by ar4** [comment]
Also, you mentioned that "the amplitude of the signal remains
consistent or even larger at the later time steps". It is certainly
possible for amplitudes to increase, depending on your velocity model,
but usually the general trend is for them to decrease. Would you like
me to check your code to make sure that it is correct?

**Comment by ruizhecao96** [comment]
Thank you very much for your help!
Here is my code:
```
import torch
import matplotlib.pyplot as plt
import deepwave
from deepwave import scalar
import numpy as np


def circle_masking(label, r, center_x, center_y):
    y, x = np.meshgrid(np.arange(label.shape[1]), np.arange(label.shape[0]))
    dist = np.sqrt((x - center_x) ** 2 + (y - center_y) ** 2)
    mask = dist <= r
    label[mask] = 1
    return label

device = torch.device('cuda')
ny = 160
nx = 200
dx = 0.2*1e-3


img = np.zeros((200, 160))
circle_masking(img, 5, 50, 20)
circle_masking(img, 5, 100, 40)
circle_masking(img, 5, 150, 120)
circle_masking(img, 5, 150, 100)
circle_masking(img, 5, 150, 140)
circle_masking(img, 5, 150, 80)
plt.imshow(img)
plt.show()

img = np.where(img > 0, 1500, 5850)

v_init = torch.from_numpy(img).to(device).float()
v = v_init.clone()
v.requires_grad_()

n_shots = 64

n_sources_per_shot = 1
d_source = 0.5*1e-3
first_source = 0
source_depth = 0

n_receivers_per_shot = 64
d_receiver = 0.5*1e-3
first_receiver = 0
receiver_depth = 0

freq = 5*1e6
fs = 25 * 1e6
nt = 750
dt = 1/fs
peak_time = 1.5 / freq

source_locations = torch.zeros(n_shots, n_sources_per_shot, 2, device=device)
source_locations[..., 0] = 0
source_locations[:, 0, 1] = torch.arange(n_shots) * d_source

receiver_locations = torch.zeros(n_shots, n_receivers_per_shot, 2, device=device)
receiver_locations[..., 0] = 0
receiver_locations[:, :, 1] = (torch.arange(n_receivers_per_shot) * d_source).repeat(n_shots, 1)

source_amplitude = (
    deepwave.wavelets.ricker(freq, nt, dt, peak_time)
    .repeat(n_shots, n_sources_per_shot, 1)
    .to(device)
)

out = deepwave.scalar(v, grid_spacing=dx, dt=dt, source_amplitudes=source_amplitude * 100, source_locations=source_locations / dx, accuracy=4,
                       receiver_locations=receiver_locations /dx, pml_width=[0, 0, 20, 20], pml_freq=freq, max_vel=5850)

fmc = out[-1].detach().cpu().numpy()
plt.plot(fmc[0, 60])
plt.show()

```

Actually, I can do the real-world experiment at the same time. While in the experiment result the amplitude is almost zero in the end step, the simulation amplitude is still quite large.

**Comment by ar4** [comment]
Thank you for the code.

You have a grid spacing of 0.2e-3 but a source and receiver spacing of
0.5e-3. Every second one will not be centred on a grid cell and so
will not be exactly where you wish it to be (Deepwave will shift it to
a grid cell centre). I suggest using a grid spacing that divides
evenly into your source/receiver spacing, or using the Hicks location
interpolation method provided with Deepwave if that is not possible.

It is recommended to have at least six (and preferably more) grid
cells per wavelength in order for the finite difference simulation to
be accurate. With a grid spacing of 0.2e-3 there are only 1.5
wavelengths of the peak frequency per grid cell in the areas with a
velocity of 1500m/s. I suggest using a smaller grid cell spacing.

Although I suspect that you will be doing inversion later, when you
are only doing forward modelling I suggest not specifying
"requires_grad" on the velocity model, as this will use substantial
extra memory.

I think the main cause of the receiver data not behaving the way that
you expected is due to the pml_width setting. In your code it is set
to make the top and bottom surfaces of the model rigid, causing strong
reflections from them. Is this what you wanted? Leaving pml_width at
the default [20, 20, 20, 20] will cause all edges to be absorbing.

**Comment by ar4** [comment]
Since you are comparing the data to a real experiment, the amplitudes
will also decay less quickly due to Deepwave being a 2D propagator.
You could try applying a transformation to the real data to make it
similar to if it had been produced by a 2D propagator, such as
https://doi.org/10.1190/1.1443798 .

**Comment by ruizhecao96** [comment]
I really appreciate these suggestions!
I do want the top and bottom to be reflective since in practice the bottom and the top will be the interface of steel and air.
I will try with your advice and eventually do the FWI, hopefully I can get a good result and keep you informed.

Best regards
Ruizhe

**Comment by l112n** [comment]
Hello，
Thank you for sharing this useful tool. I installed this toolkit before, recently came back to find that the page has changed, I remember looking for the locked file, need to enter a string of addresses, I seem to have forgotten, or can you share the previous installation method again?

**Comment by ar4** [comment]
Hello,

The old installation method is no longer required. Do you wish to find
those instructions because you need to use an old version of Deepwave? I
don't think the interface has changed, so codes written for the versions of
Deepwave that used the old installation method should work with the new
version without any changes, and the new version of Deepwave has the
benefit of being much easier to install.

**Comment by ar4** [comment]
Hello. I am going to close this Issue now, but please feel free to reopen it if you need to.
